### PR TITLE
box: add integrity check option

### PR DESCRIPTION
**Pull request by Lord-KA**
The new module can be enable only in the Enterprise Edition builds via `--integrity-check` cli option.

Needed for tarantool/tarantool-ee#585

NO_DOC=will be added to Enterprise Edition
NO_CHANGELOG=see NO_DOC
NO_TEST=see NO_DOC

### CODE REVIEW

**Review by ochaplashkin — APPROVED** [review]

**Review by Totktonada — APPROVED** [review]
I've no full picture of the activity about integrity checks, but given the code that is proposed here, I've no objections.

File: test/luatest_helpers/server.lua [hunk 1]
```diff
@@ -105,7 +105,10 @@ Server.constructor_checks = fun.chain(Server.constructor_checks, {
 function Server:initialize()
     if self.config_file ~= nil then
         self.command = arg[-1]
-        self.args = {'--name', self.alias}
+
+        self.args = self.args or {}
+        table.insert(self.args, '--name')
+        table.insert(self.args, self.alias)
```
[the remark this answers is not in the document]
Comment (Lord-KA) [reply test/luatest_helpers/server.lua]: Thanks for the suggestion! Fixed.

File: src/box/lua/config/source/file.lua [hunk 2]
```diff
@@ -69,6 +69,18 @@ function methods.sync(self, config_module, _iconfig)
     assert(config_module._config_file ~= nil)
 
     local data = universal_read(config_module._config_file, 'config file')
+
+    do
```
Comment (igormunkin) [inline src/box/lua/config/source/file.lua:73]: Minor: Could you clarify why `do ... end` is used here?

File: src/box/lua/integrity.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include "trivia/config.h"
+
+#if defined(ENABLE_INTEGRITY)
```
Comment (igormunkin) [inline src/box/lua/integrity.h:10]: Minor: Why not just use `#ifdef` here?

File: src/main.cc [hunk 4]
```diff
@@ -813,6 +818,18 @@ main(int argc, char **argv)
 		case 'F':
 			box_is_force_recovery = true;
 			break;
+		case 'I':
+#if ENABLE_INTEGRITY
+			opt_mask |= O_INTEGRITY;
+			instance.hashes = (const char *)xstrdup(optarg);
```
Comment (igormunkin) [inline src/main.cc:824]: There is no need in `xstrdup` here: it is used for `name`/`config` to make it consistent with `getenv_safe`, hence `instance.*` fields are released below. So, I suggest to just copy the pointer, but if the contents are required to be copied, please release the copied content.

File: src/trivia/config.h.cmake [hunk 5]
```diff
@@ -290,6 +290,7 @@
 #cmakedefine ENABLE_ETCD_CLIENT 1
 #cmakedefine ENABLE_CONFIG_EXTRAS 1
 #cmakedefine ENABLE_FAILOVER 1
+#cmakedefine ENABLE_INTEGRITY 1
```
Comment (igormunkin) [inline src/trivia/config.h.cmake:293]: Minor: You can use `#cmakedefine01 ENABLE_INTEGRITY` to use `#if` instead of `#ifdef`. Feel free to ignore.

**Review by igormunkin — APPROVED** [review]
**[PATCH 2/2] box: add integrity check option**

@Lord-KA, thanks for the patch! LGTM, but please consider my comments below.

**Review by igormunkin — APPROVED** [review]
**[PATCH 1/2] test: allow user-defined args in luatest_helpers.server**

@Lord-KA, thanks for the patch! LGTM.

File: src/box/lua/config/source/file.lua (hunk 2 above)
Comment (Lord-KA) [reply src/box/lua/config/source/file.lua:73]: I felt, that this code is alien to the place it was inserted, but now it seems like an overreaction, so removed.

File: src/box/lua/integrity.h (hunk 3 above)
Comment (Lord-KA) [reply src/box/lua/integrity.h:10]: I wanted to use the same style as for the other EE modules, for more details of this style see https://github.com/tarantool/tarantool/pull/8621.

File: src/main.cc (hunk 4 above)
Comment (Lord-KA) [reply src/main.cc:824]: Right, fixed that.

File: src/trivia/config.h.cmake (hunk 5 above)
Comment (Lord-KA) [reply src/trivia/config.h.cmake:293]: Ignored for consistency reasons.

File: src/box/lua/config/source/file.lua (hunk 2 above)
Comment (igormunkin) [reply src/box/lua/config/source/file.lua]: Well, honestly such splitting into blocks is widely used in Lua world, so I was just asking for your reasons... It was much better before, but if you feel it was not, I'll not insist.

File: src/box/lua/integrity.h (hunk 3 above)
Comment (igormunkin) [reply src/box/lua/integrity.h:10]: OK, thanks for clarification!

**Pull request merged**